src/main/kotlin/fi/metatavu/metaform/server/persistence/dao/ReplyDAO.kt

```diff
@@ -120,107 +120,194 @@ class ReplyDAO : AbstractDAO<Reply>() {
-    }
-    if (userId != null) {
-      restrictions.add(criteriaBuilder.equal(root.get(Reply_.userId), userId))
-    }
-    if (!includeRevisions) {
-      restrictions.add(criteriaBuilder.isNull(root.get(Reply_.revision)))
-    }
-    if (createdBefore != null) {
-      restrictions.add(criteriaBuilder.lessThanOrEqualTo(root.get(Reply_.createdAt), createdBefore))
-    }
-    if (createdAfter != null) {
-      restrictions.add(
-        criteriaBuilder.greaterThanOrEqualTo(
-          root.get(Reply_.createdAt),
-          createdAfter
-        )
+      val criteriaBuilder: CriteriaBuilder = entityManager.criteriaBuilder
+      val criteria: CriteriaQuery<Reply> = criteriaBuilder.createQuery(
+              Reply::class.java
       )
-    }
-    if (modifiedBefore != null) {
-      restrictions.add(
-        criteriaBuilder.lessThanOrEqualTo(
-          root.get(Reply_.modifiedAt),
-          modifiedBefore
-        )
+      val root = criteria.from(
+              Reply::class.java
       )
-    }
-    if (modifiedAfter != null) {
-      restrictions.add(
-        criteriaBuilder.greaterThanOrEqualTo(
-          root.get(Reply_.modifiedAt),
-          modifiedAfter
-        )
+
+      val restrictions: MutableList<Predicate> = handleReplyQueryCriteriaList(
+              criteriaBuilder = criteriaBuilder,
+              root = root,
+              metaform = metaform,
+              userId = userId,
+              includeRevisions = includeRevisions,
+              createdBefore = createdBefore,
+              createdAfter = createdAfter,
+              modifiedBefore = modifiedBefore,
+              modifiedAfter = modifiedAfter
       )
-    }
 
-    val attr = root.get(when (orderBy) {
-      ReplyOrderCriteria.CREATED -> Reply_.createdAt
-      ReplyOrderCriteria.MODIFIED -> Reply_.modifiedAt
-    })
+      val attr = root.get(when (orderBy) {
+          ReplyOrderCriteria.CREATED -> Reply_.createdAt
+          ReplyOrderCriteria.MODIFIED -> Reply_.modifiedAt
+      })
 
-    criteria.orderBy(if (latestFirst) {
-      criteriaBuilder.desc(attr)
-    } else {
-      criteriaBuilder.asc(attr)
-    })
-
-    fieldFilters?.filters?.stream()?.forEach(Consumer { fieldFilter: FieldFilter ->
-      val valuePredicate =
-        getFieldFilterValuePredicate(criteriaBuilder, criteria, root, fieldFilter)
-      if (fieldFilter.operator == FieldFilterOperator.NOT_EQUALS) {
-        restrictions.add(
-          criteriaBuilder.or(
-            valuePredicate,
-            criteriaBuilder.not(
-              criteriaBuilder.`in`(root)
-                .value(createFieldPresentQuery(criteriaBuilder, criteria, fieldFilter.field))
-            )
-          )
-        )
+      criteria.orderBy(if (latestFirst) {
+          criteriaBuilder.desc(attr)
       } else {
-        restrictions.add(valuePredicate)
+          criteriaBuilder.asc(attr)
+      })
+
+      fieldFilters?.filters?.stream()?.forEach(Consumer { fieldFilter: FieldFilter ->
```

Comment:
Wouldn't it make more sense to move also fieldFilters to handleReplyQueryCriteriaList method? It of course requires passing of criteria to the method, but I don't see any problem with that

Reply:
Moved as suggested.